### PR TITLE
add rand!, round, and rem

### LINKED ISSUE #21: `rand` does not work with Sherlog64
```
julia> using Sherlogs

julia> x=rand(Float64, 3)
3-element Vector{Float64}:
 0.40189430899764034
 0.7603918178002613
 0.5363160895268921

julia> x=rand(Sherlog64, 3)
ERROR: ArgumentError: Sampler for this object is not defined
Stacktrace:
 [1] Random.Sampler(#unused#::Type{Random.MersenneTwister}, sp::Random.SamplerTrivial{Random.CloseOpen01{Sherlog64}, Sherlog64}, #unused#::Val{1})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:145
 [2] Random.Sampler(rng::Random.MersenneTwister, x::Random.SamplerTrivial{Random.CloseOpen01{Sherlog64}, Sherlog64}, r::Val{1})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:139
 [3] rand(rng::Random.MersenneTwister, X::Random.SamplerTrivial{Random.CloseOpen01{Sherlog64}, Sherlog64})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:253
 [4] rand!
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:271 [inlined]
 [5] rand!
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:267 [inlined]
 [6] rand
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:288 [inlined]
 [7] rand
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:289 [inlined]
 [8] rand(::Type{Sherlog64}, ::Int64)
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:292
 [9] top-level scope
   @ REPL[3]:1
```

### PR DESCRIPTION
**Pull request by bjarthur**
fixes https://github.com/milankl/Sherlogs.jl/issues/21

tests currently fail b/c `promote_rule` is not working.  any ideas?

i also still need to add tests for `rand!`, etc...  and to use code-gen on DrWatson{16,32,64}.jl too.

### CODE REVIEW

File: src/SherlogN.jl [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+          :sinh, :cosh, :tanh, :csch, :sech, :coth,
+          :asinh, :acosh, :atanh, :acsch, :asech, :acoth,
+          :sinc, :sinpi, :cospi,
+          :sind, :cosd, :tand, :cscd, :secd, :cotd,
+          :asind, :acosd, :atand, :acscd, :asecd, :acotd
+         )
+    @eval function Base.$O(x::$SherlogN{T,i}) where {T,i}
+        r = $O(x.val)
+        log_it(T,r,i)
+        return Sherlog64{T,i}(r)
```
Comment (milankl) [inline src/SherlogN.jl]: This should be SherlogN instead of Sherlog64 as the result `r` has to be made a SherlogN again

File: src/SherlogN.jl [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    end
+end
+
+for O in (:(<), :(<=))
+    @eval Base.$O(x::$SherlogN, y::$SherlogN) = $O(x.val, y.val)
+end
+
+@eval function rand!(x::Array{$SherlogN})
+    for i in eachindex(x)
+        x[i] = $SherlogN.(rand($Float))
```
Comment (milankl) [inline src/SherlogN.jl]: Can we write this as `rand($SherlogN)` and define that function as something like `rand(::Type{T}) where {T<:AbstractSherlog} = convert(T,rand($Float))` ? And maybe while we're at it also define `randn`?

**Comment by milankl** [comment]
I think the `promote_rule` is fine, the error message was for promotion between `Sherlog16` and `Sherlog64` which I believe was due to line 78

File: src/SherlogN.jl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        log_it(T,r,i)
+        return $SherlogN{T,i}(r)
+    end
+end
+
+for O in (:(<), :(<=))
+    @eval Base.$O(x::$SherlogN, y::$SherlogN) = $O(x.val, y.val)
+end
+
+Random.rand(rng::AbstractRNG, ::Random.Sampler{$SherlogN}) = convert($SherlogN,rand(rng, $Float))
```
Comment (milankl) [inline src/SherlogN.jl]: Defined a `Random.Sampler` for Sherlogs. Following the `Random` documentation this is the way to implement various rand-related functions automatically for custom types. Can you check that this also works in your case?

**Comment by milankl** [comment]
Looking good now, @bjarthur can you check that this now also works in your case?

File: src/SherlogN.jl [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        log_it(T,r,i)
+        return $SherlogN{T,i}(r)
+    end
+end
+
+for O in (:(<), :(<=))
+    @eval Base.$O(x::$SherlogN, y::$SherlogN) = $O(x.val, y.val)
+end
+
+@eval Random.rand(rng::Random.AbstractRNG, ::Random.Sampler{$SherlogN}) = convert($SherlogN,rand(rng, $Float))
```
Comment (bjarthur) [inline src/SherlogN.jl]: great that defining a `Sampler` works!  i just read the [Random docs](https://docs.julialang.org/en/v1/stdlib/Random/#Generating-values-for-an-AbstractFloat-type), and they suggest to use `Random.rand(rng::Random.AbstractRNG, ::Random.SamplerTrivial{Random.CloseOpen01{T}})` instead for subtypes of `AbstractFloat`.  i tried defining that as `= convert(Sherlog16,rand(rng, Float16))` and it works, as does your method.  not sure which is better.

Comment (milankl) [reply src/SherlogN.jl]: Apparently `SamplerTrivial` is a fallback, better suited in the case where an `AbstractFloat` is wrapped into something else (as the case for Sherlogs). I guess this is more consistent, but not sure whether it makes it difference for us. Happy for another commit if you want to change it `SamplerTrivial`?

Comment (bjarthur) [reply src/SherlogN.jl]: i'm not sure `SamplerTrivial` makes a difference either. or whether `CloseOpen01{T}` is needed to exclude 1.  probably best to use it though as advised.  i also need to add tests

Comment (milankl) [reply src/SherlogN.jl]: I think it's more an implementation question regarding the function structure. `rand($Float)` can't produce 1 so I doubt a 1 would propagate through `SamplerTrivial` to yield us a `SherlogN(1.0)`

**Comment by bjarthur** [comment]
i think we're close.  just need to resolve this parametric problem:

```
julia> using Sherlogs, BFloat16s, Random
[ Info: Precompiling Sherlogs [d4954a42-a18b-4f81-bb79-24f1192c93d8]

julia> rand(Sherlog16, 3)
3-element Vector{Sherlog16}:
 Sherlog16(0.842)
 Sherlog16(0.838)
 Sherlog16(0.1699)

julia> rand(Sherlog16{BFloat16}, 3)
ERROR: ArgumentError: Sampler for this object is not defined
Stacktrace:
 [1] Random.Sampler(#unused#::Type{MersenneTwister}, sp::Random.SamplerTrivial{Random.CloseOpen01{Sherlog16{BFloat16, i} where i}, Sherlog16{BFloat16, i} where i}, #unused#::Val{1})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:145
 [2] Random.Sampler(rng::MersenneTwister, x::Random.SamplerTrivial{Random.CloseOpen01{Sherlog16{BFloat16, i} where i}, Sherlog16{BFloat16, i} where i}, r::Val{1})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:139
 [3] rand(rng::MersenneTwister, X::Random.SamplerTrivial{Random.CloseOpen01{Sherlog16{BFloat16, i} where i}, Sherlog16{BFloat16, i} where i})
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:253
 [4] rand!
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:271 [inlined]
 [5] rand!
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:267 [inlined]
 [6] rand
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:288 [inlined]
 [7] rand
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:289 [inlined]
 [8] rand(::Type{Sherlog16{BFloat16, i} where i}, ::Int64)
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/Random.jl:292
 [9] top-level scope
   @ REPL[3]:1
```

similar problem with e.g. `sin`:

```
julia> sin(Sherlog16(3))
Sherlog16(0.1411)

julia> sin(Sherlog16{BFloat16}(3))
ERROR: MethodError: no method matching BFloat16(::Float16)
Closest candidates are:
  (::Type{T})(::Real, ::RoundingMode) where T<:AbstractFloat at rounding.jl:200
  (::Type{T})(::T) where T<:Number at boot.jl:760
  (::Type{T})(::AbstractChar) where T<:Union{AbstractChar, Number} at char.jl:50
  ...
Stacktrace:
 [1] log_it(#unused#::Type{BFloat16}, x::Float16, id::Int64)
   @ Sherlogs ~/.julia/dev/Sherlogs/src/logbook.jl:41
 [2] sin(x::Sherlog16{BFloat16, 1})
   @ Sherlogs ~/.julia/dev/Sherlogs/src/SherlogN.jl:77
 [3] top-level scope
   @ REPL[8]:1
```

**Comment by bjarthur** [comment]
and `randn` doesn't work:

```
julia> rand(Sherlog32, 3)
3-element Vector{Sherlog32}:
 Sherlog32(0.6890775)
 Sherlog32(0.82779145)
 Sherlog32(0.5848477)

julia> randn(Sherlog32, 3)
ERROR: MethodError: no method matching randn(::Random.MersenneTwister, ::Type{Sherlog32})
Closest candidates are:
  randn(::Random.AbstractRNG, ::Type{T}, ::Integer, ::Integer...) where T at /groups/scicompsoft/home/arthurb/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:223
  randn(::Random.AbstractRNG, ::Type{T}, ::Tuple{Vararg{Int64, N}} where N) where T at /groups/scicompsoft/home/arthurb/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:220
  randn(::Random.AbstractRNG) at /groups/scicompsoft/home/arthurb/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:38
  ...
Stacktrace:
 [1] randn!
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:197 [inlined]
 [2] randn
   @ ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:223 [inlined]
 [3] randn(#unused#::Type{Sherlog32}, dims::Int64)
   @ Random ~/bin/julia-1.6.1/share/julia/stdlib/v1.6/Random/src/normal.jl:225
 [4] top-level scope
   @ REPL[7]:1
```

**Comment by bjarthur** [comment]
`randn` works now

**Comment by bjarthur** [comment]
added `randexp` to complete the AbstractFloat interface

**Comment by bjarthur** [comment]
all fixed now i think

**Comment by bjarthur** [comment]
should we rebase?  i see only three real commits here:  1. less code duplication, 2. add rand et al, and 3. your fixes to my sloppiness.

**Comment by milankl** [comment]
> should we rebase? 

Yes, but I also don't really mind to merge this given that it doesn't introduce breaking changes. Many thanks for all your effort!

**Comment by bjarthur** [comment]
fixup'ed down to 7 reasonably clean commits.  think it's ready to go.

one thing you might consider in future is combining SherlogsN.jl and DrWatsonN.jl with a little more code gen.  they're nearly identical.

**Comment by milankl** [comment]
Agree. But it might be fine for now, thanks for all the commits!

**Pull request merged**